pkg/mei/mkhiclient.go

```diff
@@ -0,0 +1,300 @@
+	MKHIGroupIDHmrfpo = 0x5
+	MKHIGroupIDGen    = 0xff
+)
+
+// MKHI HMRFPO command IDs, see
+// https://github.com/coreboot/coreboot/blob/b8b8ec832360ada5a313f10938bb6cfc310a11eb/src/soc/intel/common/block/include/intelblocks/cse.h#L33
+const (
+	MKHIHmrfpoEnable    = 0x1
+	MKHIHmrfpoGetStatus = 0x3
+)
+
+// pciAddressToInt converts the last part of a PCI address string to integer,
+// e.g. 0000:00:16.0 will return 0x16 .
+func pciAddressToInt(pciaddr string) (int, error) {
+	parts := strings.Split(pciaddr, ":")
+	if len(parts) != 3 {
+		return 0, fmt.Errorf("wrong PCI address string '%s': want 3 colon-separated groups, got %d", pciaddr, len(parts))
+	}
+	parts = strings.Split(parts[2], ".")
+	if len(parts) != 2 {
+		return 0, fmt.Errorf("wrong PCI address sub-component '%s': want 2 dot-separated groups, got %d", pciaddr, len(parts))
+	}
+	a, err := strconv.ParseInt(parts[0], 16, 64)
+	if err != nil {
+		return 0, fmt.Errorf("failed to parse hex string '%s': %w", parts[0], err)
+	}
+	return int(a), nil
+}
+
+// Open opens an MKHI client connection.
+func (m *MKHIClient) Open(meiPath string) error {
+	cp, err := m.MEI.Open(meiPath, MKHIGuid)
+	if err != nil {
+		return err
+	}
+	m.clientProperties = cp
+	return nil
+}
+
+// Close closes an MKHI client connection.
+func (m *MKHIClient) Close() error {
+	m.clientProperties = nil
+	return m.MEI.Close()
+}
+
+// IsHmrfpoEnableAllowed queries whether the HMRFPO enable is allowed.
+func (m MKHIClient) IsHmrfpoEnableAllowed() (bool, error) {
+	/* This is a reimplementation of cse_is_hmrfpo_enable_allowed from
+	 * coreboot/src/soc/intel/common/block/cse/cse.c . The below comment also
+	 * comes from that function.
+	 *
+	 * Allow sending HMRFPO ENABLE command only if:
+	 *  - CSE's current working state is Normal and current operation mode is Normal
+	 *  - (or) cse's current working state is normal and current operation mode is
+	 *    Soft Temp Disable if CSE's Firmware SKU is Custom
+	 *
+	 */
+	meiDev, err := GetMeiPciDevice()
+	if err != nil {
+		return false, fmt.Errorf("failed to get PCI MEI device: %w", err)
+	}
+	log.Printf("MEI Device found: %s", meiDev)
+	intelBusID := 0
+	intelMEDeviceID, err := pciAddressToInt(meiDev.Address)
+	if err != nil {
+		return false, fmt.Errorf("failed to parse MEI product ID: %w", err)
+	}
+	// usually 0. Any better way to discover this dynamically?
+	intelMEDevFn := 0
+
+	// check that CSE's current working state is normal
+	cs, err := txtsuite.PCIReadConfig32(intelBusID, int(intelMEDeviceID), intelMEDevFn, 0x40 /* PCI_ME_HFSTS1 */)
+	if err != nil {
+		return false, fmt.Errorf("PCIReadConfig32 failed: %w", err)
+	}
+	// check that the current working state is ME_HFS1_CWS_NORMAL (0x05) and
+	// current operation mode is ME_HFS1_COM_NORMAL (0x0).
+	// `working_state` is bits 1-4 and `operation_state` is bits 7-9.
+	// FIXME validate this code against the logic in this function's top
+	// comment.
+	if (cs&0xf) != 0x05 || ((cs>>6)&0x7) != 0x0 {
```

Comment:
Can you make these constants into constants?

Reply:
Sure.

Additionally I am doing bit-shifting here to avoid defining one `struct` with bit-field accessors for each check, but if you think it's a cleaner approach I can do that too

Reply:
I'd probably define the struct with booleans. Read it in with marshal/unmarshal methods. Probably?

Reply:
Any obvious downside in doing it like I did for the other `struct`s in this package? I.e. defining `type MyObject [4]byte` and having getters and setters. A bit verbose maybe, but it's a lot easier when passing data around to syscalls, or in the crazy case of I/O with unions as arguments (see ClientGUID/ClientProperties, one is for input, the other for output, same bunch of bytes)

Reply:
defined as constants - I haven't created the structured types, I checked again and they have many fields and I'd rather not do that